core/src/main/java/com/linecorp/armeria/server/metric/PrometheusExpositionService.java

```diff
@@ -85,11 +90,30 @@ public PrometheusExpositionService(CollectorRegistry collectorRegistry) {
 
     @Override
     protected HttpResponse doGet(ServiceRequestContext ctx, HttpRequest req) throws Exception {
-        final ByteArrayOutputStream stream = new ByteArrayOutputStream();
-        try (OutputStreamWriter writer = new OutputStreamWriter(stream)) {
-            TextFormat.write004(writer, collectorRegistry.metricFamilySamples());
-        }
-        return HttpResponse.of(HttpStatus.OK, CONTENT_TYPE_004, stream.toByteArray());
+        final HttpResponseWriter responseWriter = HttpResponse.streaming();
+        responseWriter.write(ResponseHeaders.builder(HttpStatus.OK).contentType(CONTENT_TYPE_004).build());
+        ctx.blockingTaskExecutor().execute(() -> {
+            try (BufferedWriter writer = new BufferedWriter(new Writer() {
+                @Override
+                public void write(char[] cbuf, int off, int len) {
+                    responseWriter.write(HttpData.ofUtf8(new String(cbuf, off, len)));
+                }
+
+                @Override
+                public void flush() {
+                }
+
+                @Override
+                public void close() {
+                    responseWriter.close();
+                }
+            }, CHAR_BUFFER_SIZE)) {
+                TextFormat.write004(writer, collectorRegistry.metricFamilySamples());
+            } catch (IOException e) {
+                responseWriter.close(e);
```

Comment:
Q) I'm wondering if the response will write the exception correctly

My guess was that if an `IOException` is thrown:
1) `responseWriter.close();` is invoked first due to `try with resources`
2) aftewards, the catch block is invoked

No problem if I'm misunderstanding and the `responseWriter.close(e);` is invoked first.
(Currently, `IOException` can't be thrown anyways IIUC 😅 )

Ref: https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html
> Note: A try-with-resources statement can have catch and finally blocks just like an ordinary try statement. In a try-with-resources statement, any catch or finally block is run after the resources declared have been closed.

Reply:
Sounds better not to use try-with-close and close the stream in the `finally` block if no exception ocurred.

Reply:
thanks, try to handle that correctly. Please check ^^